compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
That's actually a really good example: a CAS loop in an AM handler is definitely a problem, unless we can prove that the AM handler itself won't need to do anything in order for the CAS loop to eventually satisfy, and that the CAS loop will complete "pretty fast". The first of these is a definite functional requirement. The second only has performance implications, but is worth thinking hard about.  (For example, in the ugni comm layer we do a CAS loop in AM handling on Gemini for "network" AMOs that Gemini networks can't actually do and there also isn't a processor AMO. We implement these by means of AM requests to the locale that owns the target variable. We expect that such a CAS loop will only spin while some other AM handler on the same target node is simultaneously doing the same thing to the same target variable, but we expect that other AMO will complete very rapidly and we won't spin more than a time or two.)